pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +194,604 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+		Timeout: EvictionTimeSlice,
+		OnPodDeletedOrEvicted: func(pod *corev1.Pod, usingEviction bool) {
+			verbStr := "Deleted"
+			if usingEviction {
+				verbStr = "Evicted"
+			}
+			klog.Info(fmt.Sprintf("%s pod from Node", verbStr),
+				"pod", fmt.Sprintf("%s/%s", pod.Name, pod.Namespace))
+		},
+		Out:    writer{klog.Info},
+		ErrOut: writer{klog.Error},
+		DryRun: false,
+	}
+	return nil
+}
+
+func (r *ReconcileNodeMaintenance) runCordonOrUncordon(cordonOn bool, dcheck DeadlineCheck) error {
+
+	hasTaintOnNode := HasTaintOnNode(r.node, cordonOn)
+        if hasTaintOnNode != cordonOn {
+                if !dcheck.isExpired() {
+                        if err := AddOrRemoveTaint(r.drainer.Client, r.node, cordonOn); err != nil {
+                                r.reqLogger.Errorf("failed to set tain (action-set: %t) error: %v", cordonOn, err)
+                                return err
+                        }
+                        r.reqLogger.Debugf("completed add/remove taint set=%t", cordonOn)
+                } else {
+                        r.reqLogger.Info("skipped remove taint; time expired taintOn=", cordonOn)
+                }
+
+        } else {
+                r.reqLogger.Debugf("no change of taint required. set=%t", cordonOn)
+        }
+
+        if r.node.Spec.Unschedulable != cordonOn {
+                if !dcheck.isExpired() {
+                        if err := drain.RunCordonOrUncordon(r.drainer, r.node, cordonOn); err != nil {
+                                r.reqLogger.Errorf("failed to action: %t %v", cordonOn, err)
+                                return err
+                        }
+                        r.reqLogger.Debugf("completed cordon/uncordon action-set=%t", cordonOn)
+                } else {
+                        r.reqLogger.Info("skipped cordon/uncordon; time expired cordonOn=", cordonOn)
+                }
+        } else {
+                r.reqLogger.Debugf("no change of cordon required. set=%t", cordonOn)
+        }
 
 	return nil
 }
 
-func (r *ReconcileNodeMaintenance) initMaintenanceStatus(nm *kubevirtv1alpha1.NodeMaintenance) error {
-	if nm.Status.Phase == "" {
-		nm.Status.Phase = kubevirtv1alpha1.MaintenanceRunning
-		pendingList, errlist := r.drainer.GetPodsForDeletion(nm.Spec.NodeName)
-		if errlist != nil {
-			return fmt.Errorf("Failed to get pods for eviction while initializing status")
+func (r *ReconcileNodeMaintenance) evictPods(dcheck DeadlineCheck) error {
+
+	if !dcheck.isExpired() {
+		nodeName := r.node.ObjectMeta.Name
+		list, errs := r.drainer.GetPodsForDeletion(nodeName)
+		if errs != nil {
+			err := utilerrors.NewAggregate(errs)
+			r.reqLogger.Errorf("failed got get pod for eviction %v", err)
+			return err
 		}
-		if pendingList != nil {
-			nm.Status.PendingPods = GetPodNameList(pendingList.Pods())
+
+		if !dcheck.isExpired() && len(list.Pods()) != 0 {
+
+			r.drainer.Timeout = dcheck.DurationUntilExpiration()
```

Comment:
I don't think we want this

Reply:
is that better? if you set a deadline we don't want to run past it - according to your design.
[
if dcheck.isSet {
       r.drainer.Timeout = dcheck.DurationUntilExpiration()
}
](url)